src/compiler/crystal/codegen/cast.cr

```diff
@@ -204,11 +204,7 @@ class Crystal::CodeGenVisitor
     store cast_to(value, target_type), target_pointer
   end
 
-  def assign_distinct(target_pointer, target_type : VirtualMetaclassType, value_type : MetaclassType, value)
-    store value, target_pointer
-  end
-
-  def assign_distinct(target_pointer, target_type : VirtualMetaclassType, value_type : VirtualMetaclassType, value)
+  def assign_distinct(target_pointer, target_type : VirtualMetaclassType, value_type : MetaclassType | GenericClassInstanceMetaclassType | GenericModuleInstanceMetaclassType | VirtualMetaclassType, value)
```

Comment:
To clarify for following reviewers: The actual fix is the addition of `GenericClassInstanceMetaclassType` and `GenericModuleInstanceMetaclassType` as accepted types for `value_type`.
Most of the edits in this files are just refactoring and improvements to error messages.

I would have preferred two separate commits for that (maybe keep that in mind for next time 😄)